vignettes/compareToBaseR.Rmd

```diff
@@ -0,0 +1,683 @@
+
+We may want to retrieve strings that contain a pattern of interest.
+
+```{r}
+## base
+grep(pattern = "a", x = fruit, value = TRUE)
+
+
+## stringr
+str_subset(string = fruit, pattern = "a")
+```
+
+In base R we can use the value indicator in `grep()`. In stringr there is a specialized function: `str_subset()`.
+
+## `str_extract`
+
+*Extract matching patterns from a string*
+
+We may want to pick out certain patterns from a string, for example, the digits in a shopping list.
+
+```{r}
+shopping_list <- c("apples x4", "bag of flour", "bag of sugar", "milk x2")
+
+## base
+matches <- regexpr(pattern = "\\d", text = shopping_list) ## digits only
+regmatches(x = shopping_list, m = matches)
+
+gregexpr(pattern = "[a-z]+", text = shopping_list) 
+
+
+## stringr
+str_extract(string = shopping_list, pattern = "\\d") 
+str_extract_all(shopping_list, "[a-z]+")
+
+```
+
+If we want to extract matches in base R we need to use the output of `regexpr()` as input to `regmatches()` whereas in stringr we would only  need one function call to `str_extract()`. If we want to extract all matches instead of just the first, we would need to wrangle the output of base R's `gregexpr()`. In stringr, there is a specialized function to do this: `str_extract_all()`.
+
+
+## `str_match`
+
+*Extract matched groups from a string*
+
+We may also want to extract groups from a string. Here I'm going to use the scenario from Section 14.4.3 in [R for Data Science](https://r4ds.had.co.nz/strings.html) instead of the one in the documentation because I find the nouns in sentences example easier to conceptualize than the phone number example.
+
+```{r}
+head(sentences)
+noun <- "(a|the) ([^ ]+)"
+has_noun <- sentences %>%
+  str_subset(noun) %>%
+  head(5)
+has_noun
+
+## base
+matches <- regexec(pattern = noun, text = has_noun)
+do.call("rbind", regmatches(x = has_noun, m = matches))
+gregexpr(pattern = noun, text = has_noun)
+
+## stringr
+str_match(string = has_noun, pattern = noun)
+str_match_all(string = has_noun, pattern = noun)
+
+```
+
+The output of grouped matches is the full match and then the components of each group. Again, in base R we need to pair `regexec()` and `regmatches()` whereas in stringr we just need one function: `str_match()`. To get all the grouped matches in base R we would need to organize the output of `gregexpr()` whereas stringr's `str_match_all()` streamlines this process.
+
+
+# Manage lengths
+
+## `str_length`
+
+*The length of a string*
+
+To determine the length of a string, base R uses `nchar()` (not to be confused with `length()` which gives the length of vectors, etc.) while stringr uses `str_length()`.
+
+```{r}
+## base
+nchar(letters)
+
+## stringr
+str_length(letters)
+
+```
+
+There are some subtle differences between base and stringr here. `nchar()` requires a character vector, so it will return an error if used on a factor. `str_length()` can handle a factor input.
+
+
+```{r,eval=F}
```

Comment:
Better to use `error = TRUE` rather than `eval = FALSE` (then you can remove the comment recreating the error message)